pkg/lib/oauth/redis/store.go

```diff
@@ -384,6 +384,32 @@ func (s *Store) UpdateOfflineGrantApp2AppDeviceKey(grantID string, newKey string
 	return grant, nil
 }
 
+func (s *Store) UpdateOfflineGrantRefreshTokens(grantID string, refreshTokens []oauth.OfflineGrantRefreshToken, expireAt time.Time) (*oauth.OfflineGrant, error) {
```

Comment:
The mutex here does not protect concurrent write. Reader A and Reader B could read the offline grant around the same time. Reader A obtains the lock and write .RefreshTokens. Reader B then obtains the lock, and overwrite what Reader A wrote.